File: request-reply/src/broker/servidor.java
import org.zeromq.ZMQ;
import org.zeromq.ZContext;
import java.io.*;
import java.util.*;

class servidor {//Existe pois precisa saber se login deu certo ou n 
    // Dados em memória
    private static Set<String> usuariosLogados = new HashSet<>();
    private static Map<String, Long> timestampsLogin = new HashMap<>();
    
    // Arquivo próprio do servidor (salvo em volume montado)
    private static final String USERS_FILE = "/app/data/users_data.ser";
    
    public static void main(String[] args) {
        try (ZContext context = new ZContext()) {
            ZMQ.Socket socket = context.createSocket(ZMQ.REP);
            socket.connect("tcp://broker:5556");
            
            System.out.println("servidor Java Iniciado");
        
            carregarDados();
            
            while (true) {
                try {
                    // Receber por ZeroMQ
                    byte[] msgBytes = socket.recv(0);
                    
                    Message msg = MessagePackUtil.deserialize(msgBytes, Message.class);
                    
                    System.out.println("\n MENSAGEM RECEBIDA UHLLL");
                    System.out.println("Tipo: " + msg.getType());
                    System.out.println("Username: " + msg.getUsername());
                    System.out.println("Timestamp: " + msg.getTimestamp());
                    System.out.println("Bytes recebidos: " + msgBytes.length);
                    
                    Response response;
                    
                    if ("login".equals(msg.getType())) {
                        response = processarLogin(msg);
                    } else {
                        response = new Response(false, "Tipo desconhecido");
                    }
                    
                    System.out.println("\nENVIANDO RESPOSTA !!!!");
                    System.out.println("Success: " + response.isSuccess());
                    System.out.println("Message: " + response.getMessage());
                    System.out.println("Timestamp: " + response.getTimestamp());
                    
                    byte[] responseBytes = MessagePackUtil.serialize(response);
                    
                    socket.send(responseBytes, 0);
                    
                } catch (Exception e) {
                    System.err.println("Erro: " + e.getMessage());
                    Response errorResponse = new Response(false, "Erro no servidor");
                    byte[] errorBytes = MessagePackUtil.serialize(errorResponse);
                    socket.send(errorBytes, 0);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
    private static Response processarLogin(Message msg) {
        String username = msg.getUsername();
        
        if (username == null || username.trim().isEmpty()) {
            return new Response(false, "Nome de usuário inválido");
        }
        
        // Registrar
        usuariosLogados.add(username);
        timestampsLogin.put(username, msg.getTimestamp());
        
        // Persistir em disco
        salvarDados();
        
        System.out.println("Login registrado: " + username);
        return new Response(true, "Login realizado com sucesso para " + username);
    }
    
    // Salvar dados em disco
    private static void salvarDados() {
        try {
            ObjectOutputStream oos = new ObjectOutputStream(
                new FileOutputStream(USERS_FILE)
            );
            oos.writeObject(timestampsLogin);
            oos.close();
            System.out.println("Dados persistidos em disco: " + USERS_FILE);
        } catch (Exception e) {
            System.err.println(" Erro ao salvar: " + e.getMessage());
        }
    }
    
    // Carregar dados do disco
    @SuppressWarnings("unchecked")
    private static void carregarDados() {
        try {
            File file = new File(USERS_FILE);
            if (file.exists()) {
                ObjectInputStream ois = new ObjectInputStream(
                    new FileInputStream(file)
                );
                timestampsLogin = (Map<String, Long>) ois.readObject();
                usuariosLogados.addAll(timestampsLogin.keySet());
                ois.close();
                System.out.println("Dados carregados: " + usuariosLogados.size() + " usuários");
            }
        } catch (Exception e) {
            System.err.println(" Erro ao carregar: " + e.getMessage());
        }
    }
}

File: request-reply/src/broker/servidor.py
import json
import os
import zmq
import zoneinfo
from datetime import datetime
import msgpack
from dataclasses import dataclass, asdict
import pickle

@dataclass
class Message:
    type: str
    username: str
    timestamp: int
    channel_name: str = ""
    @staticmethod
    def unpack(data: bytes) -> "Message":
        d = msgpack.unpackb(data)
        return Message(**d)

@dataclass
class Response:
    success: bool
    message: str
    timestamp: int = 0
    channel_name: str=""
    def __post_init__(self):
        if self.timestamp == 0:
            self.timestamp = int(datetime.now().timestamp() * 1000)
    
    def pack(self) -> bytes:
        return msgpack.packb(asdict(self))

# ============ PERSISTÊNCIA ============
DADOS_FILE = "/app/data/server_data.pkl"

def salvar_dados():
    os.makedirs("/app/data", exist_ok=True)
    with open(DADOS_FILE, "wb") as f:  # wb = write binary
        pickle.dump({
            "usuarios": usuarios,
            "tarefas": tarefas,
            "canais": canais,
            "logins": logins
        }, f)
    print("Dados salvos.", flush=True)

def carregar_dados():
    global usuarios, tarefas, canais, logins
    if os.path.exists(DADOS_FILE):
        with open(DADOS_FILE, "rb") as f:  # rb = read binary
            dados = pickle.load(f)
        usuarios = dados.get("usuarios", [])
        tarefas  = dados.get("tarefas", [])
        canais   = dados.get("canais", [])
        logins   = dados.get("logins", [])
        print(f"Dados carregados: {len(usuarios)} usuários, {len(canais)} canais", flush=True)

# ============ SETUP ============
context = zmq.Context()
socket = context.socket(zmq.REP)
socket.connect("tcp://broker:5556")
fuso = zoneinfo.ZoneInfo("America/Sao_Paulo")


usuarios = []
logins = []
canais = []
logado = False
tempo = ""

carregar_dados()  # carrega ao iniciar

# ============ LOOP PRINCIPAL ============
while True:
    msg = Message.unpack(socket.recv())
    fazer = msg.type
    resposta = ""

    if fazer == "login":
        if msg.username in usuarios:
            tempo = msg.timestamp
            horario = datetime.now(tz=fuso).timestamp()*1000
            resposta = Response(success=True, message=f"Login realizado às {horario}")
            login = (msg.username, horario)
            logins.append(login)
            logado = True
            salvar_dados()  # persiste após login
            print(resposta.message, flush=True)
        else:
            usuarios.append(msg.username)
            logado = True
            resposta = Response(success=True, message="Cadastro inexistente. Adicionando...")
            salvar_dados()  # persiste novo usuário
            print(resposta.message, flush=True)

    elif logado == True:
        if fazer == "criar":
            tarefas.append(msg.username + "|" + str(tempo))
            resposta = Response(success=True, message="Tarefa criada")
            salvar_dados()  # persiste nova tarefa
            print(resposta.message, flush=True)

        elif fazer == "remover":
            if msg.username in tarefas:
                tarefas.remove(msg.username)
                resposta = Response(success=True, message="Tarefa removida")
                salvar_dados()  # persiste remoção
            else:
                resposta = Response(success=False, message=f"Tarefa não encontrada :(")
            print(resposta.message, flush=True)

        elif fazer == "list_channels":
            
            resposta = Response(success=True, message=f"{canais}")
        
        elif fazer == "create_channel":
            canais.append(msg.channel_name)
            resposta = Response(success=True, message="Canal Criado!!! Yipeeeeee")
            salvar_dados()


        else:
            resposta = Response(success=False, message="Comando inválido")
            print(resposta.message, flush=True)
    else:
        resposta = Response(success=False, message="Usuário não logado!")
        print(resposta.message, flush=True)

    print(f"Mensagem recebida: {msg} às {datetime.now(tz=fuso).timestamp() * 1000}", flush=True)
    socket.send(resposta.pack())